src/index.ts

```diff
@@ -771,20 +775,96 @@ class Spanner extends GrpcService {
         callback(err);
         return;
       }
-      const clientName = config.client;
+      this.getGaxClient_(config, (err, gaxClient) => {
+        if (err) {
+          callback(err);
+          return;
+        }
+        let reqOpts = extend(true, {}, config.reqOpts);
+        reqOpts = replaceProjectIdToken(reqOpts, projectId!);
+        const requestFn = gaxClient[config.method].bind(
+          gaxClient,
+          reqOpts,
+          config.gaxOpts
+        );
+        callback(null, requestFn);
+      });
+    });
+  }
+
+  /**
+   * get GAX client. This will retrive the end point uris for specific instance and cache the client accordingly.
+   *
+   * @private
+   *
+   * @param {string} projectId Request config
+   * @param {object} config Request config
+   * @param {function} callback Callback function
+   */
+  getGaxClient_(config, callback) {
+    let clientName = config.client;
+    if (
+      clientName !== 'SpannerClient' ||
+      !this.options.enableResourceBasedRouting
+    ) {
       if (!this.clients_.has(clientName)) {
-        this.clients_.set(clientName, new gapic.v1[clientName](this.options));
+        this.setSpannerClient(clientName, config, this.options);
       }
       const gaxClient = this.clients_.get(clientName)!;
-      let reqOpts = extend(true, {}, config.reqOpts);
-      reqOpts = replaceProjectIdToken(reqOpts, projectId!);
-      const requestFn = gaxClient[config.method].bind(
-        gaxClient,
-        reqOpts,
-        config.gaxOpts
+      callback(null, gaxClient);
+      return;
+    }
+
+    if (!config.instanceId) {
+      const error = new Error('instanceId is required.');
+      callback(error);
+      return;
+    }
+    const instanceId = config.instanceId;
+    clientName = `${clientName}-${instanceId}`;
+
+    if (!this.clients_.has(clientName)) {
+      this.instances_.get(instanceId)!.getMetadata(
+        {
+          fieldNames: ['endpointUris'],
+        },
+        (err, instance) => {
+          if (err) {
+            if (err.code === 7) {
+              process.emitWarning(
+                'spanner.instances.get permission must be added to use resource based routing.'
```

Comment:
Sorry we didn't quite update the spec on time but the warning message should be the following:

"The client library attempted to connect to an endpoint closer to your Cloud Spanner data but was unable to do so. The client library will fallback to the API endpoint given in the client options, which may result in increased latency. We recommend including the scope https://www.googleapis.com/auth/spanner.admin so that the client library can get an instance-specific endpoint and efficiently route requests."

Reply:
@skuruppu Looks like the link is broken or not working, cloud you look into that.

Reply:
In https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.admin.instance.v1 this is how the scopes are named (scroll down to GetInstance).

Reply:
Done!